homeassistant/components/color_extractor/__init__.py

```diff
@@ -0,0 +1,146 @@
+"""Module for color_extractor (RGB extraction from images) component."""
+import asyncio
+import io
+import logging
+
+from PIL import UnidentifiedImageError
+import aiohttp
+import async_timeout
+from colorthief import ColorThief
+import voluptuous as vol
+
+from homeassistant.components.color_extractor.const import (
+    ATTR_PATH,
+    ATTR_URL,
+    DOMAIN,
+    SERVICE_PREDOMINANT_COLOR_FILE,
+    SERVICE_PREDOMINANT_COLOR_URL,
+)
+from homeassistant.components.light import (
+    ATTR_RGB_COLOR,
+    DOMAIN as LIGHT_DOMAIN,
+    LIGHT_TURN_ON_SCHEMA,
+    SERVICE_TURN_ON,
+)
+from homeassistant.helpers import aiohttp_client
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+# Extend the existing light.turn_on service schema
+SERVICE_SCHEMA = cv.make_entity_service_schema(
+    {
+        **LIGHT_TURN_ON_SCHEMA,
+        vol.Exclusive(ATTR_PATH, "color_extractor"): cv.isfile,
+        vol.Exclusive(ATTR_URL, "color_extractor"): cv.url,
+    }
+)
+
+
+def _get_file(file_path):
+    """Get a PIL acceptable input file reference.
+
+    Allows us to mock patch during testing to make BytesIO stream.
+    """
+    return file_path
+
+
+async def async_setup(hass, hass_config):
+    """Set up services for color_extractor integration."""
+
+    def _get_color(file_handler) -> tuple:
+        """Given an image file, extract the predominant color from it."""
+        try:
+            color_thief = ColorThief(file_handler)
+        except UnidentifiedImageError as ex:
+            _LOGGER.error("Bad image file provided, are you sure it's an image? %s", ex)
```

Comment:
I think that it is better to catch this inside the service handling, so the error message can contain the url or path that didn't work

Reply:
Moving the exception handling out to each service call, I can remove the `Optional[tuple]` as it's either tuple or an exception.